openupgradelib/migrator.py

```diff
@@ -0,0 +1,549 @@
+    return wrapper
+
+
+class Migrator(object):
+    def __init__(self, cr, addon, table_names=None, context=None):
+        """Start a migrator instance for a given addon.
+
+        :param openerp.sql_db.Cursor cr:
+            Database cursor to use for the migration.
+
+        :param str addon:
+            This is the name of the addon where data is attached before running
+            the migration.
+
+        :param dict table_names:
+            Dictionary mapping model names to table names, such as::
+
+                {"res.partner": "res_partner"}
+
+            Only useful when a model's table **is not** the model's name
+            replacing dots by low lines **and** the model name cannot be found
+            in Odoo's regitry anymore (maybe because you are deleting it).
+
+        :param dict context:
+            Default context for the environment created, in case you need it.
+        """
+        self.env = api.Environment(cr, SUPERUSER_ID, context or dict())
+        self.addon = addon
+        self.table_names = table_names or None
+
+    @contextmanager
+    def _allow_pgcodes(self, *codes):
+        """Shortcut for :meth:`~.allow_pgcodes` that uses instanced cursor."""
+        return allow_pgcodes(self.env.cr, *codes)
+
+    def _execute(self, query, params=None, log_exceptions=True):
+        """Wrapper that logs before executing.
+
+        Parameters are the same as for :meth:`~.Cursor.execute`, except that
+        :param:`log_exceptions` defaults to ``True``.
+        """
+        _logger.info("Executing query with params %s:\n%s", params, query)
+        self.env.cr.execute(query, params, log_exceptions)
+
+    def _fetch_ids(self):
+        """Return list of ids from last executed query."""
+        return tuple(row["id"] for row in self.env.cr.dictfetchall())
+
+    def _model_id(self, model_name):
+        """Get a model's ID in ``ir.model`` table.
+
+        :param str model_name:
+            The model name formatted like ``res.partner``.
+        """
+        self._execute(
+            "SELECT id FROM ir_model WHERE model = %s",
+            (model_name,))
+        return self.env.cr.fetchone()[0]
+
+    def _record_remove(self, model_name, ids, xmlid=True):
+        """Remove a record from the database.
+
+        :param str model_name:
+            Model name from where to remove the record. E.g. ``res.partner``.
+
+        :param int ids:
+            Tuple of IDs of the record to remove.
+
+        :param bool xmlid:
+            Wether to remove the XMLID for this record in this addon too.
+        """
+        table = self._table_name(model_name)
+        self._execute("DELETE FROM %s WHERE id IN %%s" % table, (ids,))
+        if xmlid:
+            self._execute(
+                """ DELETE FROM ir_model_data
+                    WHERE module = %s AND model = %s AND res_id IN %s
+                """,
+                (self.addon, model_name, ids))
+
+    @wip
+    def _table_fk(self, table_name):
```

Comment:
@StefanRijnhart My comment was actually meant for the `def _table_fk(self, table_name):` function. Sorry, github doesn't display as I intended. 

@Yajo I think the your migrator can be analyzed one by one as I might guess there are overlapping features. I was pleased to see `_table_fk(self, table_name)` and it would be nice if a `copy_table` or `rename_table` would in some distant future be aware of 2many relational tables.